packages/instruction-plans/src/instruction-plan.ts

```diff
@@ -1 +1,526 @@
+ * - {@link SequentialInstructionPlan} - A plan that contains other plans that
+ *   must be executed sequentially. It also defines whether the plan is divisible
+ *   meaning that instructions inside it can be split into separate transactions.
+ * - {@link MessagePackerInstructionPlan} - A plan that can dynamically fill
+ *  instructions into transaction messages.
+ *
+ * Helpers are provided for each of these plans to make it easier to create them.
+ *
+ * @example
+ * ```ts
+ * const myInstructionPlan: InstructionPlan = parallelInstructionPlan([
+ *    sequentialInstructionPlan([instructionA, instructionB]),
+ *    instructionC,
+ *    instructionD,
+ * ]);
+ * ```
+ *
+ * @see {@link SingleInstructionPlan}
+ * @see {@link ParallelInstructionPlan}
+ * @see {@link SequentialInstructionPlan}
+ * @see {@link MessagePackerInstructionPlan}
+ */
+export type InstructionPlan =
+    | MessagePackerInstructionPlan
+    | ParallelInstructionPlan
+    | SequentialInstructionPlan
+    | SingleInstructionPlan;
+
+/**
+ * A plan wrapping other plans that must be executed sequentially.
+ *
+ * It also defines whether nested plans are divisible — meaning that
+ * the instructions inside them can be split into separate transactions.
+ * When `divisible` is `false`, the instructions inside the plan should
+ * all be executed atomicly — either in a single transaction or in a
+ * transaction bundle.
+ *
+ * You may use the {@link sequentialInstructionPlan} and {@link nonDivisibleSequentialInstructionPlan}
+ * helpers to create objects of this type.
+ *
+ * @example
+ * Simple sequential plan with two instructions.
+ * ```ts
+ * const plan = sequentialInstructionPlan([instructionA, instructionB]);
+ * plan satisfies SequentialInstructionPlan;
+ * ```
+ *
+ * @example
+ * Non-divisible sequential plan with two instructions.
+ * ```ts
+ * const plan = nonDivisibleSequentialInstructionPlan([instructionA, instructionB]);
+ * plan satisfies SequentialInstructionPlan & { divisible: false };
+ * ```
+ *
+ * @example
+ * Sequential plan with nested parallel plans.
+ * Here, instructions A and B can be executed in parallel, but they must both be finalized
+ * before instructions C and D can be sent — which can also be executed in parallel.
+ * ```ts
+ * const plan = sequentialInstructionPlan([
+ *   parallelInstructionPlan([instructionA, instructionB]),
+ *   parallelInstructionPlan([instructionC, instructionD]),
+ * ]);
+ * plan satisfies SequentialInstructionPlan & { divisible: false };
+ * ```
+ *
+ * @see {@link sequentialInstructionPlan}
+ * @see {@link nonDivisibleSequentialInstructionPlan}
+ */
+export type SequentialInstructionPlan = Readonly<{
+    divisible: boolean;
+    kind: 'sequential';
+    plans: InstructionPlan[];
+}>;
+
+/**
+ * A plan wrapping other plans that can be executed in parallel.
+ *
+ * This means direct children of this plan can be executed in separate
+ * parallel transactions without causing any side effects.
```

Comment:
It's not so much without side effects as it is without consequences. I can still run a bunch of transactions in parallel that produce side effects, but so long as the ordering of those side effects are not consequential to each other, it's safe to run them in parallel.
```suggestion
 * parallel transactions without consequence.
```